src/ferc_xbrl_extractor/instance.py

```diff
@@ -0,0 +1,255 @@
+"""Parse a single instance."""
+import io
+from enum import Enum, auto
+from typing import BinaryIO, Dict, List, Optional, Union
+
+from lxml import etree  # nosec: B410
+from lxml.etree import _Element as Element  # nosec: B410
+from pydantic import BaseModel, validator
+
+XBRL_INSTANCE = "http://www.xbrl.org/2003/instance"
+
+
+class Instance:
+    """Class to manage parsing XBRL filings."""
+
+    def __init__(self, file_info: Union[str, BinaryIO], name: str):
+        """
+        Construct Instance class.
+
+        Args:
+            file_info: Either path to filing, or file data.
+            name: Name of filing.
+        """
+        self.name = name
+        self.file = file_info
+
+    def parse(self, fact_prefix: str = "ferc"):
+        """
+        Parse a single XBRL instance using XML library directly.
+
+        Args:
+            fact_prefix: Prefix to identify facts in filing (defaults to 'ferc').
+
+        Returns:
+            context_dict: Dictionary of contexts in filing.
+            fact_dict: Dictionary of facts in filing.
+            filing_name: Name of filing.
+        """
+        # Create parser to enable parsing 'huge' xml files
+        parser = etree.XMLParser(huge_tree=True)
+
+        # Check if instance contains path to file or file data and parse accordingly
+        if isinstance(self.file, str):
+            tree = etree.parse(self.file, parser=parser)  # nosec: B320
+            root = tree.getroot()
+        elif isinstance(self.file, io.BytesIO):
+            root = etree.fromstring(self.file.read(), parser=parser)  # nosec: B320
+        else:
+            raise TypeError("Can only parse XBRL from path to file or file like object")
+
+        context_dict = {}
+        fact_dict = {}
+
+        contexts = root.findall(f"{{{XBRL_INSTANCE}}}context")
+        facts = root.findall(f"{fact_prefix}:*", root.nsmap)
+
+        for context in contexts:
+            new_context = Context.from_xml(context)
+            context_dict[new_context.c_id] = new_context
+            fact_dict[new_context.c_id] = []
+
+        for fact in facts:
+            new_fact = Fact.from_xml(fact)
+            if new_fact.value is not None:
+                fact_dict[new_fact.c_id].append(new_fact)
+
+        return context_dict, fact_dict, self.name
```

Comment:
Does it make sense to create a class to hold contexts, dicts and name?

Reply:
Yeah I've been struggling to come up with the best way to handle this. I definitely don't like passing around all these dictionaries though. I think I have a refactor in mind that will clean this up, but I'm leaning towards holding off until after this review is closed. The refactor would touch several parts of the code base, so I don't want it to drag this review out too much, but let me know what you think?